// sjcl-bit-array/index.js
/** @fileOverview Arrays of bits, encoded as arrays of Numbers.
 *
 * @author Emily Stark
 * @author Mike Hamburg
 * @author Dan Boneh
 */

/** @namespace Arrays of bits, encoded as arrays of Numbers.
 *
 * @description
 * <p>
 * These objects are the currency accepted by SJCL's crypto functions.
 * </p>
 *
 * <p>
 * Most of our crypto primitives operate on arrays of 4-byte words internally,
 * but many of them can take arguments that are not a multiple of 4 bytes.
 * This library encodes arrays of bits (whose size need not be a multiple of 8
 * bits) as arrays of 32-bit words.  The bits are packed, big-endian, into an
 * array of words, 32 bits at a time.  Since the words are double-precision
 * floating point numbers, they fit some extra data.  We use this (in a private,
 * possibly-changing manner) to encode the number of bits actually  present
 * in the last word of the array.
 * </p>
 *
 * <p>
 * Because bitwise ops clear this out-of-band data, these arrays can be passed
 * to ciphers like AES which want arrays of words.
 * </p>
 */

module.exports = {
  bitSlice: require('./bit-slice'),
  extract: require('./extract'),
  concat: require('./concat'),
  bitLength: require('./bit-length'),
  clamp: require('./clamp'),
  partial: require('./partial'),
  getPartial: require('./get-partial'),
  equal: require('./equal'),
  _shiftRight: require('./shift-right'),
  _xor4: require('./xor4'),
  byteswapM: require('./byteswap-m') 
};
